Modify note content about kubectl, CRDs, and subresources (#34468)
kubernetes/website: content/en/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch.md

And you can see the `retainKeys` strategy in the
[Kubernetes API documentation](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#deploymentspec-v1-apps).

> ## Use subresource flag with kubectl patch command to update the deployment

Reviewer: I would move this inside:
```markdown
## Alternate forms of the kubectl patch command
```

Also, a text suggestion:

```suggestion
## Update an object's replica count using `kubectl patch` with `--subresource` {#scale-kubectl-patch}

{{< feature-state for_k8s_version="v1.24" state="alpha" >}}
```
(is this right, alpha in v1.24?)

Author: Yes, it is alpha in v1.24.
Ref: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cli/2590-kubectl-subresource/kep.yaml#L24

Author: addressed

> In kubernetes v1.24, kubectl commands like get, patch, edit and replace will now contain a new flag `--subresource=[subresource-name]` which will allow fetching and updating `status` and `scale` subresources for all resources (built-in and CRs) that support these subresources.

Reviewer: I'm not happy with this text. It sounds like it was written before Kubernetes v1.24 was released.
Instead, please describe how the latest Kubernetes release works, right now. Explain that this is generic across any API with a `status` or `scale` subresource, and that the Deployment is an example.

> Here's the configuration file for a Deployment that has two replicas.

Reviewer: (nit)
```suggestion
Here's a manifest for a Deployment that has two replicas:
```

Author: addressed

{{< codenew file="application/deployment-subresource.yaml" >}}

Create the Deployment:

```shell
kubectl apply -f https://k8s.io/examples/application/deployment-subresource.yaml
```

View the Pods associated with your Deployment:

```shell
kubectl get pods
```

> In the output, you can see that Deployment has two Pods.

Reviewer: ```suggestion
In the output, you can see that Deployment has two Pods. For example:
```

Author: addressed

```
NAME                                READY   STATUS    RESTARTS   AGE
subresource-demo-777d58f9d8-7wzff   1/1     Running   0          28s
subresource-demo-777d58f9d8-fkf4l   1/1     Running   0          28s
```

> Now, Patch your Deployment with `--subresource=[subresource-name]` flag:

Reviewer: ```suggestion
Now, patch that Deployment with `--subresource=[subresource-name]` flag:
```

```shell
kubectl patch deployment subresource-demo --subresource='scale' --type='merge' -p '{"spec":{"replicas":3}}'
```

> you can see following output, which means your deployment is successfully patched:

Reviewer: ```suggestion
The output is:
```

```shell
scale.autoscaling/subresource-demo patched
```

View the Pods associated with your patched Deployment:

```shell

> kubectl get pods

Reviewer: Use a label selector here to only show the demo Pods.

```

In the output, you can see one new pod is created, so now you have 3 running pods.

```
NAME                                READY   STATUS    RESTARTS   AGE
subresource-demo-777d58f9d8-7wzff   1/1     Running   0          2m10s
subresource-demo-777d58f9d8-fkf4l   1/1     Running   0          2m10s
subresource-demo-777d58f9d8-v8cjk   1/1     Running   0          19s
```

View the patched Deployment:

```shell
kubectl get deployment subresource-demo -o yaml
```

```yaml

> spec:

Reviewer: ```suggestion
...
spec:
```
You could also use jsonpath output to select just the status.

Author: we can use jsonpath but on using it give output in the below format:

```
$ kubectl get deployment subresource-demo  -o=jsonpath='{.status}'
{"availableReplicas":3,"conditions":[{"lastTransitionTime":"2022-08-17T07:35:17Z","lastUpdateTime":"2022-08-17T07:35:57Z","message":"ReplicaSet \"subresource-demo-777d58f9d8\" has successfully progressed.","reason":"NewReplicaSetAvailable","status":"True","type":"Progressing"},{"lastTransitionTime":"2022-08-18T08:52:25Z","lastUpdateTime":"2022-08-18T08:52:25Z","message":"Deployment has minimum availability.","reason":"MinimumReplicasAvailable","status":"True","type":"Available"}],"observedGeneration":2,"readyReplicas":3,"replicas":3,"updatedReplicas":3}
```

IMO it seems good and also  YAML content is used to show the output in the corresponding docs, so for the uniformity of the docs, we can write in this way.

  replicas: 3
  ...
status:
  availableReplicas: 3
  readyReplicas: 3
  replicas: 3
```

{{< note >}}

> If `--subresource` flag is used for a resource that doesn't support the subresource, a NotFound error will be returned.

Reviewer: ```suggestion
If you run `kubectl patch` and specify `--subresource` flag for resource that doesn't support that
particular subresource, the API server returns a 404 Not Found error.
```

Author: addressed

kubectl patch deployment patch-demo --patch '{"spec": {"template": {"spec": {"containers": [{"name": "patch-demo-ctr-2","image": "redis"}]}}}}'
```

### Update an object's replica count using `kubectl patch` with `--subresource` {#scale-kubectl-patch}

{{< feature-state for_k8s_version="v1.24" state="alpha" >}}

> The flag `--subresource=[subresource-name]` is used with kubectl commands like get, patch, edit and replace to fetch and update `status` and `scale` subresources of the resources (applicable for kubectl version v1.24 or more). This flag is used with all the API resources (built-in and CRs) which has `status` or `scale` subresource, Deployment is one of the examples which supports these subresources.

Reviewer 2: ```suggestion
The flag `--subresource=[subresource-name]` is used with kubectl commands like get, patch, edit and replace to fetch and update `status` and `scale` subresources of the resources (applicable for kubectl version v1.24 or more). This flag is used with all the API resources (built-in and CRs) which has `status` or `scale` subresource. Deployment is one of the examples which supports these subresources.
```

> The flag `--subresource=[subresource-name]` is used with kubectl commands like get, patch, edit and replace to fetch and update `status` and `scale` subresources of the resources (applicable for kubectl version v1.24 or more). This flag is used with all the API resources (built-in and CRs) which has `status` or `scale` subresource. Deployment is one of the examples which supports these subresources.

Reviewer 3: PLEASE WRAP THE LONG LINES, e.g.

```suggestion
The flag `--subresource=[subresource-name]` is used with kubectl commands like get, patch,
edit and replace to fetch and update `status` and `scale` subresources of the resources
(applicable for kubectl version v1.24 or more). This flag is used with all the API resources
(built-in and CRs) which has `status` or `scale` subresource. Deployment is one of the examples
which supports these subresources.
```

Author: I will update this as per your comment soon.

Author: addressed

> The flag `--subresource=[subresource-name]` is used with kubectl commands like get, patch, edit and replace to fetch and update `status` and `scale` subresources of the resources (applicable for kubectl version v1.24 or more). This flag is used with all the API resources (built-in and CRs) which has `status` or `scale` subresource, Deployment is one of the examples which supports these subresources.

Author: addressed

Here's a manifest for a Deployment that has two replicas:

> {{< codenew file="application/deployment-subresource.yaml" >}}

Reviewer 3: Do we have to introduce a new YAML for this task?

Author: Hi @tengqm, thanks for taking look at this.

>Do we have to introduce a new YAML for this task?

For every example/strategy used [In Update API Objects in Place Using kubectl patch](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/) task, there is a separate deployment YAML and has different configurations according to strategy. So for this, we can use a new deployment YAML that is related to this only so it will be more useful to the users/readers. it maintains the uniformity of the docs.

Reviewer 3: @Shubham82 I asked this question because I failed to find anything special in this `deployment-subresource.yaml` file. In fact, it can be any generic Deployment, right? This was not meant to be a blocker for this PR. My intent is to make sure that we are adding new example files for good reasons. We have [a test case for all examples](https://github.com/kubernetes/website/blob/main/content/en/examples/examples_test.go). I'm not convinced that a new file should be added for this scenario. Thoughts?

Author: Hi @tengqm
I agree with your point, it is a generic Deployment, but we can't use the deployment mentioned in the task. if the same (kind of) generic deployment is already used in the docs we can use that one also. could you please point me to that deployment I will test and update the content accordingly to that? WDYT?

Reviewer 3: Here are some existing Deployment manifests:

- content/en/examples/application/deployment.yaml
- content/en/examples/application/deployment-scale.yaml
- content/en/examples/application/simple_deployment.yaml 
- content/en/examples/application/update_deployment.yaml 

They are pretty much the same thing repeated again and again.

Author: I have updated it. I used content/en/examples/application/deployment.yaml manifest for this particular task.
PTAL
